File: app/src/main/java/n7/ad2/items/ItemsPagedListAdapter.java
package n7.ad2.items;

import android.arch.paging.PagedListAdapter;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import n7.ad2.R;
import n7.ad2.databinding.ItemListItemBinding;
import n7.ad2.items.db.ItemModel;

public class ItemsPagedListAdapter extends PagedListAdapter<ItemModel, ItemsPagedListAdapter.ViewHolder> {

    private static final DiffUtil.ItemCallback<ItemModel> DIFF_CALLBACK = new DiffUtil.ItemCallback<ItemModel>() {
        @Override
        public boolean areItemsTheSame(@NonNull ItemModel items, @NonNull ItemModel t1) {
            return items.getCodeName().equals(t1.getCodeName());
        }

        @Override
        public boolean areContentsTheSame(@NonNull ItemModel items, @NonNull ItemModel t1) {
            return true;
        }
    };

    private LayoutInflater inflater;
    private ItemsFragment fragment;

    ItemsPagedListAdapter(ItemsFragment fragment) {
        super(DIFF_CALLBACK);
        this.fragment = fragment;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (inflater == null) inflater = LayoutInflater.from(viewGroup.getContext());
        ItemListItemBinding binding = DataBindingUtil.inflate(inflater, R.layout.item_list_item, viewGroup, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int i) {
        ItemModel item = getItem(i);
        if (item != null) {
            viewHolder.bindTo(item);
        } else {
            viewHolder.clear();
        }
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        ItemListItemBinding binding;

        ViewHolder(ItemListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        private void bindTo(ItemModel item) {
            binding.setItem(item);
            binding.setFragment(fragment);
            binding.executePendingBindings();
        }

        private void clear() {
            binding.ivItemListItem.setImageResource(R.drawable.item_placeholder);
            binding.tvItemListItem.setText("");
        }
    }


}

File: app/src/main/java/n7/ad2/ui/items/ItemsPagedListAdapter.kt
package n7.ad2.ui.items

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import coil.clear
import n7.ad2.R
import n7.ad2.databinding.ItemItemBodyBinding
import n7.ad2.databinding.ItemItemHeaderBinding
import n7.ad2.ui.items.domain.vo.VOItem
import n7.ad2.ui.items.domain.vo.VOItemBody
import n7.ad2.ui.items.domain.vo.VOItemHeader

class ItemsPagedListAdapter(
    fragment: ItemsFragment,
) : PagedListAdapter<VOItem, RecyclerView.ViewHolder>(DiffCallback()) {

    companion object {
        const val SPAN_SIZE_ITEM = 1
        const val SPAN_SIZE_ITEM_HEADER = 4
    }

    private val itemClickListener = { model: ItemItemBodyBinding -> fragment.startItemInfoFragment(model.model!!, model) }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder = when (viewType) {
        R.layout.item_item_header -> HeaderViewHolder.from(parent)
        else -> ItemViewHolder.from(parent, itemClickListener)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = getItem(position)
        when (holder) {
            is ItemViewHolder -> if (item != null) holder.bindTo(item as VOItemBody) else holder.clear()
            is HeaderViewHolder -> if (item != null) holder.bindTo(item as VOItemHeader)
        }
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        super.onViewRecycled(holder)
        when (holder) {
            is ItemViewHolder -> holder.clear()
            is HeaderViewHolder -> holder.clear()
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when (getItem(position)) {
            is VOItemBody -> R.layout.item_item_body
            is VOItemHeader -> R.layout.item_item_header
            else -> super.getItemViewType(position)
        }
    }

    private class HeaderViewHolder private constructor(
        private val binding: ItemItemHeaderBinding,
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bindTo(model: VOItemHeader) = binding.let {
            it.model = model
            it.executePendingBindings()
        }

        fun clear() = Unit

        companion object {
            fun from(parent: ViewGroup): HeaderViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ItemItemHeaderBinding.inflate(layoutInflater, parent, false)
                return HeaderViewHolder(binding)
            }
        }

    }

    private class ItemViewHolder private constructor(
        private val binding: ItemItemBodyBinding,
        private val itemClickListener: (ItemItemBodyBinding) -> Unit,
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bindTo(model: VOItemBody) = binding.let {
            it.model = model
            it.itemClickListener = itemClickListener
            it.binding = binding
            it.executePendingBindings()
        }

        fun clear() = binding.apply {
            iv.clear()
            tv.text = ""
        }

        companion object {
            fun from(parent: ViewGroup, itemClickListener: (ItemItemBodyBinding) -> Unit): ItemViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ItemItemBodyBinding.inflate(layoutInflater, parent, false)
                return ItemViewHolder(binding, itemClickListener)
            }
        }

    }

    private class DiffCallback : DiffUtil.ItemCallback<VOItem>() {
        override fun areItemsTheSame(oldItem: VOItem, newItem: VOItem) = when (oldItem) {
            is VOItemBody -> newItem is VOItemBody
            is VOItemHeader -> newItem is VOItemHeader
        }

        override fun areContentsTheSame(oldItem: VOItem, newItem: VOItem) = oldItem == newItem

    }

}